### PR TITLE
build(deps): Upgrade dependencies

### PR DESCRIPTION
**Pull request by chitralverma**
### Which issue does this PR close?

Closes https://github.com/roapi/roapi/issues/263
Related PR https://github.com/roapi/roapi/pull/264

### What changes are included in this PR?
 
This PR upgrades various dependencies (see list below) of roapi to the latest supported versions.
- datafusion (`14` -> `23`)
- arrow (`26` -> `37`)
- object_store (`0.5.4` -> `0.5.6`)
- convergence (`internal patch` -> `0.11.0` latest tag)
- sqlparser (`0.27` -> `0.33`)
- delta-rs (`patched 0.5.0` -> `patched 0.10.0`)

Other than this, this PR includes 
 - changes for the deprecated `ObjectStoreProvider`,
 - changes due to datafusion upgrade,
 - changes for `&str` -> `String` to fix lifetime issues.

### Are there any user-facing changes?

Mostly these are internal upgrades, so user-facing changes should be none.

### Point of consideration for reviewers

- Seems like the upgrades to datafusion have changed the serialization of timestamp types which required modification in test cases ([see changes](https://github.com/roapi/roapi/pull/275/files#diff-beae34fcfbf1e8b79ece90fa19ffe3840df783d6d6b51f124f79d0d13e786978)). 
- the `Cargo.toml` explicitly includes the `object_store` dependency which is currently coming from both `delta-rs` and `datafusion`. So do you think removing this and using it transitively instead makes sense?

### CODE REVIEW

**Comment by chitralverma** [comment]
@houqp @jychen7 please check this out. Thanks.

**Comment by chitralverma** [comment]
the breaks in the test cases are due to the older arrow 26 coming from connector-x which is conflicting with the new arrow 37.

raised https://github.com/sfu-db/connector-x/issues/493

**Comment by houqp** [comment]
> Seems like the upgrades to datafusion have changed the serialization of timestamp types which required modification in test cases ([see changes](https://github.com/roapi/roapi/pull/275/files#diff-beae34fcfbf1e8b79ece90fa19ffe3840df783d6d6b51f124f79d0d13e786978)).

This serialization change looks good to me :+1: 

> the Cargo.toml explicitly includes the object_store dependency which is currently coming from both delta-rs and datafusion. So do you think removing this and using it transitively instead makes sense?

Yes, we can use the export from one of the crates, probably makes more sense to use the one exported from datafusion.

File: Cargo.toml [hunk 1]
```diff
@@ -8,7 +8,3 @@ split-debuginfo = "unpacked"
 [profile.release]
 lto = true
 codegen-units = 1
-
-[patch.crates-io]
```
Comment (houqp) [inline Cargo.toml:12]: @chitralverma we can patch connectorx to use our own fork at https://github.com/roapi/connector-x if it's taking a long time for upstream to merge your PR.

**Comment by rtyler** [comment]
FYI, I just cut deltalake 0.10.0 which has the datafusion 23 and arrow 37 dependencies referenced, roapi might be able to move away from git sha1s now? :)

**Comment by chitralverma** [comment]
> FYI, I just cut deltalake 0.10.0 which has the datafusion 23 and arrow 37 dependencies referenced, roapi might be able to move away from git sha1s now? :) 

Awesome, I was waiting for this 😁

File: Cargo.toml (hunk 1 above)
Comment (chitralverma) [reply Cargo.toml:12]: Sounds good, can you please sync that fork with the latest upstream

**Comment by chitralverma** [comment]
> Yes, we can use the export from one of the crates, probably makes more sense to use the one exported from datafusion.

@houqp I think we can do the same for `sqlparser` as well.

File: Cargo.toml (hunk 1 above)
Comment (houqp) [reply Cargo.toml:12]: Just sent you an invite for write access to the repo ;)

**Review by houqp — APPROVED** [review]

File: columnq/Cargo.toml [hunk 2]
```diff
@@ -13,18 +13,18 @@ path = "src/lib.rs"
 [dependencies]
 # pulling arrow-schema manually to enable the serde feature.
 # TODO: add serde feature in datafusion to avoid this workaround
-arrow-schema = { version ="26", features = ["serde"] }
```
Comment (houqp) [inline columnq/Cargo.toml:16]: @chitralverma we might need to set the arrow-schema version to 38 so it remains the same as the version pinned in datafusion to fix the build issue, see https://github.com/apache/arrow-datafusion/commit/fc5d67aa12ec11dc43f61f1381db1dd54dcef784#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542L51

**Comment by houqp** [comment]
Thanks @chitralverma again for the patch, I pushed a fix based off your branch with connectorx upgrade in https://github.com/roapi/roapi/pull/279

**Comment by chitralverma** [comment]
> ushed a fix based off your branch with co

Thanks @houqp

**Pull request closed without merging**